### ISSUE TITLE
In a 2 server farm scenario SPProductUpdate wrongly evaluates that the patch has been already installed. 

### ISSUE THREAD

**Issue by MarekSarad** [PowerShell/SharePointDsc#839]
In a 2 server farm scenario SPProductUpdate wrongly evaluates that August 2017 CULD (Language Dependant)  patch has been already installed. 

        SPProductUpdate InstallCU
            {
                SetupFile            = 'C:\Temp\'+$CUFile 
                ShutdownServices     = $true
                Ensure               = "Present"
                PsDscRunAsCredential = $SetupAccount
            }

        # Reboot if pending

        xPendingReboot RebootCheck1
        {
            Name = "RebootCheck1"
        }

        SPProductUpdate InstallCUL
            {
                SetupFile            = 'C:\Temp\'+$CULDFile 
                ShutdownServices     = $true
                Ensure               = "Present"
                PsDscRunAsCredential = $SetupAccount
                DependsOn = "[File]CUDirectoryCopy"
            }

        xPendingReboot RebootCheck2
        {
            Name = "RebootCheck2"
        }
        
        SPConfigWizard PSConfig
        {
            Ensure = "Present"
            PsDscRunAscredential = $SetupAccount
        }

Windows Server 2012 R2 

Sharepoint 2016

SharepointDSC 2.2.0.0

I have hit the issue with the CULD on the August 2017 CU

Assumptions
	- SP 2016 has 2 types of patches. Language independent (CU) and Language dependent (CULD) 
	- Both need to be installed even if no language pack has been installed as English is as well an build in Language Pack. 
	- Ref: https://blogs.technet.microsoft.com/stefan_gossner/2017/08/08/august-2017-cu-for-sharepoint-server-2016-is-available-for-download/

So have a 2 server farm scenario (1 APP, 1 FE) applied same DSC to both. 

THE CU that I try to apply is the Aug 2017 CU and CULD (Language dependent) 

I have end up in this situation

APP: CU and CULD applied
FE: CU only applied. 

When I try to reapply DSC and get to the  SPProductUpdate InstallCUL (see above for the DSC) it tells me that patch is already installed. 

(I added some verbose logging) 

VERBOSE: [DFS-SP16FE-750]: [[SPProductUpdate]InstallCUL] The lowest version of any SharePoint component is 16.0.4573.1000 | July 11th 2018 14:37:34 Info

VERBOSE: [DFS-SP16FE-750]: [[SPProductUpdate]InstallCUL] Version of SharePoint is equal or greater than the patch version. Patch is installed July 11th 2018 14:37:34 Info

Usually CU and CULD will have same version number 
But that is not the case with Aug 17 CU for the SP2016 here CULD has lower number.  

CU: sts2016-kb4011049-fullfile-x64-glb.exe - 16.0.4573.1002
CULD: wssloc2016-kb4011053-fullfile-x64-glb.exe - 16.0.4573.1000

Running the Get-SPDscFarmVersionInfo gives me this information. 

Name                           Value                                                                                                                                             
----                           -----                                                                                                                                             
Highest                        16.0.4573.1002                                                                                                                                    
Lowest                         16.0.4573.1000 

So this fragment of the SPProductUpdate.psm1 will return false and resource will think that the update is already installed. 

 Write-Verbose -Message "The lowest version of any SharePoint component is $($versionInfo.Lowest)"
    if ($versionInfo.Lowest -lt $fileVersion)

I haven't hit that issue during previous runs of the testing (with CUs from previous months) So it might be the case that in particular case when CU and CULD are not of the same version SPProductUpdate if failing to make correct evaluation that the patch is required. 

Executing from SharepointDSc.Util.psm1

$productVersions.GetServerProductInfo($server.id) gives us clear information that on FE this patch is missing. 

ServerId                         : 10cf73ce-d986-459b-a56e-c1192def3032
ServerName                       : <Removed>
Products                         : {Microsoft SharePoint Server 2016}
RequiredButMissingProducts       : {}
OptionalMissingProducts          : {}
RequiredButMissingPatchableUnits : {}
RequiredButMissingPatches        : {Update for Microsoft SharePoint Enterprise Server 2016 (**KB4011053**) 64-Bit Edition}
InstallStatus                    : InstallRequired

When I install the CULD from GUI it installs normally.

**Comment by ykuijs** [comment]
Hi @MarekSarad, we have been discussing a different way of detecting if a patch has been deployed or not. However this is a quite tricky process. You can check out the discussion in issue #679.

At this moment @andikrueger is working on implementing a new method.

**Comment by MarekSarad** [comment]
Yes I have seen that thread and looking forward to it, the method currently used in DSC is having some issues. on Sept17 CU I have hit another issue. Similar case but for some reason the Util Get-SPDscFarmVersionInfo is not bringing results for the FE. 

On APP I get: 

PS C:\Users\_spinstall> Get-SPDscFarmVersionInfo

Name                           Value                                                                                                                  
----                           -----                                                                                                                  
Highest                        16.0.4588.1001                                                                                                         
Lowest                         16.0.4588.1000

And on FE I get:

PS C:\Windows\system32> Get-SPDscFarmVersionInfo
Get-SPServer : The server named 'DFS-SP16FE-750' does not exist.
At C:\Program Files\WindowsPowerShell\Modules\SharePointDSC\2.2.0.0\modules\SharePointDsc.Util\SharePointDsc.Util.psm1:161 char:15
+     $server = Get-SPServer -Identity $env:COMPUTERNAME
+               ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : InvalidArgument: (Microsoft.Share...PServerPipeBind:SPServerPipeBind) [Get-SPServer], PSArgumentException
    + FullyQualifiedErrorId : Microsoft.SharePoint.PowerShell.SPCmdletGetServer
 
Cannot convert argument "serverIdGuid", with value: "", for "GetServerProductInfo" to type "System.Guid": "Cannot convert null to type "System.Guid"."
At C:\Program Files\WindowsPowerShell\Modules\SharePointDSC\2.2.0.0\modules\SharePointDsc.Util\SharePointDsc.Util.psm1:166 char:5
+     $serverProductInfo = $productVersions.GetServerProductInfo($serve ...
+     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (:) [], MethodException
    + FullyQualifiedErrorId : MethodArgumentConversionInvalidCastArgument
 

Name                           Value                                                                                                                       
----                           -----                                                                                                                       
Lowest                                                                                                                                                     
Highest

Obviously the server exists but get-spfarm is bringing empty result for server property when executed from that server:

Executed from APP:

```
PS C:\Users\_spinstall> $farm = Get-SPFarm

PS C:\Users\_spinstall> $farm.servers

Address                        Role                           Status              
-------                        ----                           ------              
dfs-smtp.dfs.un.org            Invalid                        Online              
DFS-SP16AP-751                 ApplicationWithSearch          Online              
DFS-SP16FE-750                 WebFrontEndWithDistributedC... Online              
SP2016INTDB                    Invalid                        Online 
```

Executed from FE:

```
Name                           Value                                                                                                                       
----                           -----                                                                                                                       
Lowest                                                                                                                                                     
Highest                                                                                                                                                    

PS C:\Windows\system32> $farm = Get-SPFarm

PS C:\Windows\system32> $farm.servers

PS C:\Windows\system32> 
```

**Comment by johlju** [comment]
I labeled this as in progress, since work is being done in issue #679. If issue #679 does not solve this issue then we should remove 'in progress' label and switch it to 'help wanted'.

**Comment by ykuijs** [comment]
@MarekSarad This looks like you have an issue with your SharePoint environment unrelated to SharePointDsc. If you can't run the (Get-SPFarm).Servers command from PowerShell, something else is happening in your environment.

**Comment by ykuijs** [comment]
Major overhaul of the ProductUpdate resource will be merged via PR #1046

**Comment by ykuijs** [comment]
Closed via PR #1046
